src/controller/proxy/controller.go

```diff
@@ -0,0 +1,230 @@
+	"github.com/goharbor/harbor/src/replication/registry"
+	serror "github.com/goharbor/harbor/src/server/error"
+	"github.com/opencontainers/go-digest"
+	"io"
+	"net/http"
+	"sync"
+	"time"
+)
+
+const (
+	maxWait             = 10
+	manifestListWaitSec = 1800
+	sleepIntervalSec    = 20
+)
+
+var (
+	// Ctl is a global proxy controller instance
+	Ctl = NewController()
+)
+
+// Controller defines the operations related with pull through proxy
+type Controller interface {
+	// UseLocalManifest check if the manifest should use local
+	UseLocalManifest(ctx context.Context, p *models.Project, art lib.ArtifactInfo) bool
+	// UseLocalBlob check if the blob should use local
+	UseLocalBlob(ctx context.Context, p *models.Project, digest string) bool
+	// ProxyBlob proxy the blob request to the target server
+	ProxyBlob(ctx context.Context, p *models.Project, repo string, dig string, w http.ResponseWriter, r RemoteInterface) error
+	// ProxyManifest proxy the manifest to the target server
+	ProxyManifest(ctx context.Context, p *models.Project, repo string, art lib.ArtifactInfo, w http.ResponseWriter, r RemoteInterface) error
+}
+
+type controller struct {
+	blobCtl     blob.Controller
+	registryMgr registry.Manager
+	mu          sync.Mutex
+	inflight    map[string]interface{}
+	artifactCtl artifact.Controller
+	local       LocalInterface
+}
+
+func (c *controller) isProxyReady(p *models.Project) bool {
+	if p.RegistryID < 1 {
+		return false
+	}
+	reg, err := c.registryMgr.Get(p.RegistryID)
+	if err != nil {
+		log.Errorf("failed to get registry, error:%v", err)
+		return false
+	}
+	return reg.Status == model.Healthy
+}
+
+func (c *controller) UseLocalManifest(ctx context.Context, p *models.Project, art lib.ArtifactInfo) bool {
+	if !c.isProxyReady(p) {
+		return true
+	}
+	if len(string(art.Digest)) > 0 {
+		exist, err := c.local.BlobExist(ctx, art.Digest)
+		if err == nil && exist {
+			return true
+		}
+	}
+	return false
+}
+
+func (c *controller) UseLocalBlob(ctx context.Context, p *models.Project, digest string) bool {
+	if !c.isProxyReady(p) {
+		return true
+	}
+	exist, err := c.local.BlobExist(ctx, digest)
+	if err != nil {
+		return false
+	}
+	return exist
+}
+
+func (c *controller) ProxyManifest(ctx context.Context, p *models.Project, repo string, art lib.ArtifactInfo, w http.ResponseWriter, r RemoteInterface) error {
+	var man distribution.Manifest
+	var err error
+	if len(string(art.Digest)) > 0 {
+		// pull by digest
+		log.Debugf("Getting manifest by digiest %v", art.Digest)
+		man, err = r.ManifestByDigest(repo, string(art.Digest))
+	} else { // pull by tag
+		if len(art.Tag) == 0 {
+			art.Tag = "latest"
+		}
+		log.Debugf("Getting manifest by tag %v", art.Tag)
+		man, err = r.ManifestByTag(repo, string(art.Tag))
+	}
+
+	if err != nil {
+		if errors.IsNotFoundErr(err) && len(art.Tag) > 0 {
+			go func() {
+				c.local.DeleteManifest(ctx, repo, string(art.Tag))
+			}()
+		}
+		serror.SendError(w, err)
```

Comment:
If we write error to the client directly here, is necessary to return error for this function anymore?